File: 1914-cyclically-rotating-a-grid/solution.py
class Solution:
    def rotateGrid(self, grid: list[list[int]], k: int) -> list[list[int]]:
        m, n = len(grid), len(grid[0])
        num_layers = min(m, n) // 2
        
        for layer in range(num_layers):
            top, left = layer, layer
            bottom, right = m - 1 - layer, n - 1 - layer
            
            # 1. Extract the current layer elements
            elements = []
            
            # Top row: left to right (excluding last)
            for j in range(left, right):
                elements.append(grid[top][j])
            # Right col: top to bottom (excluding last)
            for i in range(top, bottom):
                elements.append(grid[i][right])
            # Bottom row: right to left (excluding last)
            for j in range(right, left, -1):
                elements.append(grid[bottom][j])
            # Left col: bottom to top (excluding last)
            for i in range(bottom, top, -1):
                elements.append(grid[i][left])
            
            # 2. Calculate effective rotation
            L = len(elements)
            shift = k % L
            
            # Rotation: index i in result list is index (i + shift) % L in original
            # Instead of creating a new list, we can just use a pointer
            idx = 0
            
            # 3. Fill the grid back with rotated elements
            # Top row
            for j in range(left, right):
                grid[top][j] = elements[(idx + shift) % L]
                idx += 1
            # Right col
            for i in range(top, bottom):
                grid[i][right] = elements[(idx + shift) % L]
                idx += 1
            # Bottom row
            for j in range(right, left, -1):
                grid[bottom][j] = elements[(idx + shift) % L]
                idx += 1
            # Left col
            for i in range(bottom, top, -1):
                grid[i][left] = elements[(idx + shift) % L]
                idx += 1
                
        return grid

File: 1914-cyclically-rotating-a-grid/solution.cpp
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<vector<int>> rotateGrid(vector<vector<int>>& grid, int k) {
        int m = grid.size();
        int n = grid[0].size();
        int layers = min(m, n) / 2;

        for (int layer = 0; layer < layers; ++layer) {
            int top = layer, left = layer;
            int bottom = m - 1 - layer, right = n - 1 - layer;

            // 1. Extract the current layer into a 1D vector
            vector<int> elements;
            // Top row
            for (int j = left; j < right; ++j) elements.push_back(grid[top][j]);
            // Right col
            for (int i = top; i < bottom; ++i) elements.push_back(grid[i][right]);
            // Bottom row
            for (int j = right; j > left; --j) elements.push_back(grid[bottom][j]);
            // Left col
            for (int i = bottom; i > top; --i) elements.push_back(grid[i][left]);

            // 2. Determine effective rotation
            int len = elements.size();
            int shift = k % len;

            // 3. Put elements back into the grid after rotation
            // The element at index 'idx' in the rotated ring 
            // comes from index (idx + shift) % len in the original ring.
            int idx = 0;
            auto get_next_val = [&]() {
                return elements[(idx++ + shift) % len];
            };

            for (int j = left; j < right; ++j) grid[top][j] = get_next_val();
            for (int i = top; i < bottom; ++i) grid[i][right] = get_next_val();
            for (int j = right; j > left; --j) grid[bottom][j] = get_next_val();
            for (int i = bottom; i > top; --i) grid[i][left] = get_next_val();
        }

        return grid;
    }
};